Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -154,31 +159,117 @@ namespace ROS2
         {
             sensor_msgs::msg::PointField pf;
-            pf.name = point_field_names[i];
+            pf.name = pointFieldNames[i];
             pf.offset = i * 4;
             pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
             pf.count = 1;
             message.fields.push_back(pf);
         }
 
-        size_t sizeInBytes = lastScanResults.m_points.size() * sizeof(AZ::Vector3);
+        if (lastScanResults.m_ids.has_value())
+        {
+            sensor_msgs::msg::PointField pfId;
+            pfId.name = "entity_id";
+            pfId.offset = message.point_step;
+            pfId.datatype = sensor_msgs::msg::PointField::INT32;
+            pfId.count = 1;
+            message.fields.push_back(pfId);
+            constexpr auto fieldLength = sizeof(int32_t);
+            message.point_step += fieldLength;
+        }
+        AZStd::array<AZ::Color, LidarSensorConfiguration::maxClass> colorLookupTable;
+        const bool publishSegmentation = lastScanResults.m_classes.has_value() && m_lidarCore.m_lidarConfiguration.m_isSegmentationEnabled;
+        if (publishSegmentation)
+        {
+            colorLookupTable = m_lidarCore.m_lidarConfiguration.GenerateSegmentationColorsLookupTable();
+
+            AZStd::array<const char*, 3> colorFieldNames = { "r", "g", "b" };
+
+            for (int i = 0; i < colorFieldNames.size(); i++)
+            {
+                sensor_msgs::msg::PointField pf;
+                pf.name = colorFieldNames[i];
+                pf.offset = message.point_step;
+                pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
+                pf.count = 1;
+                message.fields.push_back(pf);
+                message.point_step += sizeof(float);
+            }
+
+            sensor_msgs::msg::PointField pfClassId;
+            pfClassId.name = "class_id";
+            pfClassId.offset = message.point_step;
+            pfClassId.datatype = sensor_msgs::msg::PointField::UINT8;
+            pfClassId.count = 1;
+            message.fields.push_back(pfClassId);
+            message.point_step += sizeof(uint8_t);
+        }
+
+        const auto sizeInBytes = pointCount * message.point_step;
         message.data.resize(sizeInBytes);
+
+        float x, y, z;
+        for (int i = 0; i < pointCount; ++i)
+        {
+            // to avoid alignment issues, we copy the data field by field
+            x = lastScanResults.m_points[i].GetX();
+            y = lastScanResults.m_points[i].GetY();
+            z = lastScanResults.m_points[i].GetZ();
+
+            memcpy(&message.data[i * message.point_step], &x, sizeof(float));
+            memcpy(&message.data[i * message.point_step + sizeof(float)], &y, sizeof(float));
+            memcpy(&message.data[i * message.point_step + 2 * sizeof(float)], &z, sizeof(float));
+            uint32_t nextFieldOffset = 3 * sizeof(float);
+
+            if (lastScanResults.m_ids.has_value())
```

Comment:
I would suggest to avoid branching in a hot loop. You could separate this into a separate loop if possible.

Reply:
Solved with redesign to lambdas, I'm assuming compiler will handle this via magic of inlining. Also this kind of branching isn't expensive as branch like this will be correctly predicted ~100% times by branch predictor.